dcos/marathon.py

```diff
@@ -45,105 +47,154 @@ def _get_marathon_url(toml_config):
+                status_code=e.response.status_code,
+                reason=e.response.reason,
+                request_method=e.response.request.method,
+                request_url=e.response.request.url,
+                json_body=json_body)
+            raise DCOSException(message)
+
+
+ERROR_JSON_SCHEMA = {
+    'type': 'object',
+    'properties': {
+        'errors': {
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'error': {
+                        'type': 'string'
+                    }
+                },
+                'required': ['error']
+            }
+        },
+        'message': {
+            'type': 'string'
+        }
+    },
+    'anyOf': [
+        {'required': ['errors']},
+        {'required': ['message']}
+    ]
+}
+
+_ERROR_JSON_VALIDATOR = jsonschema.Draft4Validator(ERROR_JSON_SCHEMA)
+
+
+def response_error_message(
+        status_code, reason, request_method, request_url, json_body):
+    """Renders a human-readable error message from the given response data.
+
+    :param status_code: the integer status code from an HTTP response
+    :type status_code: int
+    :param reason: textual (human-readable) representation of the status code
+    :type reason: str
+    :param request_method: the HTTP method used for the request
+    :type request_method: str
+    :param request_url: the URL the request was sent to
+    :type request_url: str
+    :param json_body: the response body, parsed as JSON, or None if parsing
+                      failed
+    :type json_body: dict | list | str | int | bool | None
+    :return: the rendered error message
+    :rtype: str
     """
-    try:
-        return fn(*args, **kwargs)
-    except DCOSHTTPException as e:
-        raise _to_exception(e.response)
+
+    if status_code == 400:
+        template = 'Error on request [{} {}]: HTTP 400: {}{}'
+        suffix = ''
+        if json_body is not None:
+            json_str = json.dumps(json_body, indent=2, sort_keys=True)
+            suffix = ':\n' + json_str
+        return template.format(request_method, request_url, reason, suffix)
+
+    if status_code == 409:
+        return ('App, group, or pod is locked by one or more deployments. '
+                'Override with --force.')
+
+    if json_body is not None:
+        if not _ERROR_JSON_VALIDATOR.is_valid(json_body):
+            logger.error('Marathon server did not return a message: %s',
+                         json_body)
+
+            return _default_marathon_error()
+
+        message = json_body.get('message')
+        if message is not None:
+            return 'Error: {}'.format(message)
+
+        message = '\n'.join(err['error'] for err in json_body['errors'])
+        return _default_marathon_error(message)
+
+    template = 'Error decoding response from [{}]: HTTP {}: {}'
+    return template.format(request_url, status_code, reason)
```

Comment:
I kept the wording of most of the error messages the same as before (although I did mention pods in the message for status 409). Once the pods integration tests are working against a Marathon instance, I might look at whether the error messaging can be made more specific.